### PR TITLE
Save mode merge / Jdbc schema evolution

### PR DESCRIPTION
**Pull request by zzeekk**
### What changes are included in the pull request?
- implement SDLSaveMode.Merge for DeltaLakeTableDataObject
- implement SDLSaveMode.Merge, auto-create table and schema evolution for JdbcTableDataObject
- extend CanWriteDataFrame with forceSaveMode, so that Actions can override saveMode set in DataObject configuration
- Implement merge mode for DeduplicateAction (#235)
- implement Schema Evolution for JdbcTableDataObject (#216)

### Why are the changes needed?
see #235, #216

### CODE REVIEW

File: docs/Reference.md [hunk 1]
```diff
@@ -386,6 +386,15 @@ How it works: under the hood a PySpark DataFrame is a proxy for a Java Spark Dat
 
 ## Schema Evolution
 SmartDataLakeBuilder is built to support schema evolution where possible. This means that data pipelines adapt themselves automatically to additional or removed columns and changes of data types if possible.
+The following cases can be distinguished:
+* Overwrite all (CopyAction): if all data of a DataObject is overwritten, the schema can replaced: additional columns are added, removed columns are removed and data types are changed. Requirements:
```
Comment (pgruetter) [inline docs/Reference.md:390]:
```suggestion
* Overwrite all (CopyAction): if all data of a DataObject is overwritten, the schema can be replaced: additional columns are added, removed columns are removed and data types are changed. Requirements: 
```

File: docs/Reference.md [hunk 2]
```diff
@@ -386,6 +386,15 @@ How it works: under the hood a PySpark DataFrame is a proxy for a Java Spark Dat
...
+* Overwrite all (CopyAction): if all data of a DataObject is overwritten, the schema can replaced: additional columns are added, removed columns are removed and data types are changed. Requirements: 
+  * Output DataObject needs to be able to replace schema.
+* Overwrite all keeping existing data (Historize- & DeduplicateAction): Action consolidates new data with existing data. The schema needs to be evolved: additional columns are added with null value for existing records, removed columns are kept with null values for new records and data types are changed to new data type if supported. Requirements: 
+  * Output DataObject needs to be able to replace schema.
+  * Output DataObject must be a TransactionalSparkTableDataObject (read existing data and overwrite new data in the same SparkJob, preventing data loss in case of errors).
+* Overwrite incremental using merge (CopyAction, DeduplicateAction): Action incrementally merges new data int existing data. The schema needs to be evolved: additional columns are added with null value for existing records, removed columns are kept with null values for new records and data types are changed to new data type if supported. Requirements:
```
Comment (pgruetter) [inline docs/Reference.md:395]:
```suggestion
* Overwrite incremental using merge (CopyAction, DeduplicateAction): Action incrementally merges new data into existing data. The schema needs to be evolved: additional columns are added with null value for existing records, removed columns are kept with null values for new records and data types are changed to new data type if supported. Requirements:
```

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala [hunk 3]
```diff
@@ -143,75 +143,160 @@ case class JdbcTableDataObject(override val id: DataObjectId,
...
+   * Deleted columns will remain in the table and are made nullable.
+   */
+  private def evolveTableSchema(newSchemaRaw: StructType)(implicit session: SparkSession, context: ActionPipelineContext): Unit = {
+    val existingSchema = getExistingSchema.get
+    val newSchema = if (SchemaUtil.isSparkCaseSensitive) newSchemaRaw else StructType(SchemaUtil.prepareSchemaForDiff(newSchemaRaw, ignoreNullable = false, caseSensitive = false))
+    // prepare changes
+    val newColumns = newSchema.fieldNames.diff(existingSchema.fieldNames) // add new column
+    val missingNotNullColumns = existingSchema.fieldNames.diff(newSchema.fieldNames) // make missing columns nullable
+      .filter { col =>
+        // as Spark doesn't now if a field is nullable in the database, but we can check jdbc metadata
```
Comment (pgruetter) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala:168]:
```suggestion
        // as Spark doesn't know if a field is nullable in the database, but we can check jdbc metadata
```

**Review by pgruetter — APPROVED** [review]
Sorry for the delay. Still couldn't check all the code but it looks good to me. Just added a few comments.
Awesome features! :+1: :+1:

**Comment by zzeekk** [comment]
You're right, that was much code - just realized it now.

**Pull request merged**